File: Conversor/PHP/Cli/tests/ConversorTest.php
<?php

require_once __DIR__ . '/../conversor.php';

$pass = 0;
$fail = 0;

function test(string $name, callable $fn): void {
    global $pass, $fail;
    try {
        $fn();
        $pass++;
        echo "PASS: $name\n";
    } catch (Throwable $e) {
        $fail++;
        fwrite(STDERR, "FAIL: $name - " . $e->getMessage() . "\n");
    }
}

test('lengthConversion', function () {
    assert(abs(convert(1, "m", "cm") - 100) < 0.001);
});

test('weightConversion', function () {
    assert(abs(convert(1, "kg", "g") - 1000) < 0.001);
});

test('temperatureCtoF', function () {
    assert(abs(convert(0, "C", "F") - 32) < 0.001);
});

test('temperatureCtoK', function () {
    assert(abs(convert(0, "C", "K") - 273.15) < 0.001);
});

test('temperatureFtoC', function () {
    assert(abs(convert(32, "F", "C") - 0) < 0.001);
});

test('temperatureFtoK', function () {
    assert(abs(convert(32, "F", "K") - 273.15) < 0.001);
});

test('temperatureKtoC', function () {
    assert(abs(convert(273.15, "K", "C") - 0) < 0.001);
});

test('temperatureKtoF', function () {
    assert(abs(convert(273.15, "K", "F") - 32) < 0.001);
});

test('invalidUnit', function () {
    $thrown = false;
    try {
        convert(1, "m", "kg");
    } catch (InvalidArgumentException $e) {
        $thrown = true;
    }
    assert($thrown);
});

test('incompatibleCategories', function () {
    $thrown = false;
    try {
        convert(1, "m", "kg");
    } catch (InvalidArgumentException $e) {
        $thrown = true;
    }
    assert($thrown);
});

test('listCategories', function () {
    $cats = list_categories();
    assert(in_array("length", $cats, true));
    assert(in_array("weight", $cats, true));
    assert(in_array("temperature", $cats, true));
});

test('kmToMi', function () {
    assert(abs(convert(1, "km", "mi") - 0.621371) < 0.001);
});

test('lbToOz', function () {
    assert(abs(convert(1, "lb", "oz") - 16) < 0.001);
});

test('identity', function () {
    assert(abs(convert(100, "cm", "cm") - 100) < 0.001);
});

if ($fail > 0) {
    fwrite(STDERR, "{$fail} test(s) failed\n");
    exit(1);
}
echo "OK: {$pass} tests passed\n";

File: Conversor/Java/Cli/src/test/java/com/portfolio/conversor/ConversorTest.java
package com.portfolio.conversor;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ConversorTest {

    @Test
    void lengthConversion() {
        assertEquals(100, Conversor.convert(1, "m", "cm"), 1e-9);
        assertEquals(0.621371, Conversor.convert(1, "km", "mi"), 1e-9);
        assertEquals(1, Conversor.convert(100, "cm", "m"), 1e-9);
    }

    @Test
    void weightConversion() {
        assertEquals(1000, Conversor.convert(1, "kg", "g"), 1e-9);
        assertEquals(16, Conversor.convert(1, "lb", "oz"), 0.01);
        assertEquals(1, Conversor.convert(1000, "g", "kg"), 1e-9);
    }

    @Test
    void temperatureCtoF() {
        assertEquals(32, Conversor.convert(0, "C", "F"), 1e-9);
        assertEquals(212, Conversor.convert(100, "C", "F"), 1e-9);
    }

    @Test
    void temperatureCtoK() {
        assertEquals(273.15, Conversor.convert(0, "C", "K"), 1e-9);
    }

    @Test
    void temperatureFtoC() {
        assertEquals(0, Conversor.convert(32, "F", "C"), 1e-9);
    }

    @Test
    void temperatureFtoK() {
        assertEquals(273.15, Conversor.convert(32, "F", "K"), 1e-9);
    }

    @Test
    void temperatureKtoC() {
        assertEquals(0, Conversor.convert(273.15, "K", "C"), 1e-9);
    }

    @Test
    void temperatureKtoF() {
        assertEquals(32, Conversor.convert(273.15, "K", "F"), 1e-9);
    }

    @Test
    void temperatureIdentity() {
        assertEquals(100, Conversor.convert(100, "C", "C"), 1e-9);
    }

    @Test
    void incompatibleUnitsThrows() {
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                () -> Conversor.convert(1, "m", "kg"));
        assertEquals("Incompatible units: m -> kg", ex.getMessage());
    }

    @Test
    void incompatibleTemperatureUnitsThrows() {
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                () -> Conversor.convert(1, "C", "m"));
        assertEquals("Incompatible units: C -> m", ex.getMessage());
    }

    @Test
    void categoriesListed() {
        String[] cats = Conversor.categories();
        assertEquals(3, cats.length);
        assertEquals("length", cats[0]);
        assertEquals("weight", cats[1]);
        assertEquals("temperature", cats[2]);
    }

    @Test
    void unitsPerCategory() {
        String[] length = Conversor.unitsFor("length");
        assertEquals(6, length.length);
        assertEquals("m", length[0]);
        assertEquals("cm", length[5]);
        String[] temperature = Conversor.unitsFor("temperature");
        assertEquals(3, temperature.length);
        assertEquals("K", temperature[2]);
    }
}
